### ISSUE TITLE
[@typescript-eslint/unbound-method] Why is rule triggered in this example

### ISSUE THREAD

**Issue by doberkofler** [typescript-eslint/typescript-eslint#522]
I cannot seem to understand why the following reduced example generates a warning?

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/unbound-method": "warn"
  }
}
```

```TS
console.log(Boolean(document.addEventListener));
```

**Expected Result**
no warning

**Actual Result**
```
warning: Avoid referencing unbound methods which may cause unintentional scoping of `this` (@typescript-eslint/unbound-method) at x.ts:1:21:
> 1 | console.log(Boolean(document.addEventListener));
    |                     ^
  2 |
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `12.2.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
You are passing an object's method by value, meaning the `this` context could get redefined, breaking the method.
The fact that that particular method doesn't care about the `this` context is irrelevant to the rule's function.

Check out the docs for more information:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/unbound-method.md

An example:
![image](https://user-images.githubusercontent.com/7462525/57640598-ce9d3000-7567-11e9-8182-eaa8e457acba.png)

**Comment by mysticatea** [comment]
But this issue case, it looks just a cast. The `Boolean(xxx)` converts `xxx` to a boolean value. I wonder it's nice if the rule excepts the call of the wrapper classes of primitives.

**Comment by bradzacher** [comment]
We could potentially do that.

I'm not sure if we should be adding special cases to rules for things like this.
Esp when this specific case has an easy solution in user land:
```ts
document.addEventListener != null
```

Is it worth adding extra code and making assumptions about things?
I don't know!

**Comment by felixfbecker** [comment]
This makes the rule unusable right now, getting errors like this in a lot of places:

![image](https://user-images.githubusercontent.com/10532611/62417747-54f17b80-b658-11e9-9597-5264a1df6cdc.png)

Pretty sure TSLint did not flag these because I only see them now that I'm trying to migrate to ESLint

**Comment by bradzacher** [comment]
@felixfbecker - feel free to open a new bug for this.
this thread is about an unrelated issue.

**Comment by felixfbecker** [comment]
How is it unrelated? To me it seems to be the same issue - referencing a method in a boolean context, to e.g. check that it is not `undefined`, triggers an error (but didn't in TSLint)

**Comment by bradzacher** [comment]
Because this particular issue isn't about a boolean context, it's about allowing it within certain call expressions that are known to be good ahead of time (i.e. the Boolean function). This is a feature request.

Your issue is entirely separate - it is specifically about allowing unbound methods to be used in a unary expression, which obviously should be allowed, and thus would be considered a bug.

**Comment by ab-pm** [comment]
Similar (and probably more common) use case:
```js
// in a Jest test:
expect(mockInstance.someMethod).toHaveBeenCalled();
```

**Comment by jsaguet** [comment]
Indeed this rule does not seem adapted to jest tests

**Comment by JoshuaKGoldberg** [comment]
@bradzacher what do you think about adding an allowlist akin to what's being discussed in https://github.com/typescript-eslint/typescript-eslint/pull/4436 that defaults to the built-in primitive classes?

**Comment by bradzacher** [comment]
[For the jest case - `eslint-plugin-jest` has an extension rule which is built to support jest syntax](https://github.com/jest-community/eslint-plugin-jest/blob/HEAD/docs/rules/unbound-method.md) which is intended to be used specifically in test files and removes false-positives for jest usecases.

----

> what do you think about adding an allowlist

I dislike it, personally. Allowlists are always contextless and can thus lead to false negatives really, really easily.

For example:
```ts
function Boolean(cb: () => boolean) {
  return cb();
}

class Foo {
  state = false;
  myMethod() {
    return this.state;
  }
}

Boolean(new Foo().myMethod); // oops - false negative!
```

#4436 is really there to work around two unavoidable limitations: (1) TS lacks the expressiveness to declare a true "deep readonly" type and (2) TS's type information does not expose enough information to allow us to fully determine if a type is deeply readonly. I don't like an allowlist for it - but there I don't have any other viable options - the user can't just change their code to fix it, so they ***need*** a suppression comment to bypass the false-positive.

In this case - as mentioned the jest case has a solution in the form of the library owning an extension rule which teaches it safe cases within the rule. The other case of `Boolean` - well it can easily be worked around by changing the code to a logical check instead of a boolean-ish coercion.

---

I was going to say that this case could also be handled by just allowing unbound methods to be passed in an `any` location. However both [`Boolean`](https://github.com/microsoft/TypeScript/blob/ff3b458714056cbbe26ee4c14029162e4f185504/lib/lib.es5.d.ts#L536) and [jest's `expect`](https://github.com/microsoft/TypeScript/blob/ff3b458714056cbbe26ee4c14029162e4f185504/lib/lib.es5.d.ts#L536) are actually loosely generically typed - so there's not really any way we can statically verify that these functions are "safe" or any way for us to just ignore the cases.

I don't have a good solution! But given that this issue has 5 participants and no additional thumbs - it probably isn't a wide-spread issue?

**Comment by JoshuaKGoldberg** [comment]
Agreed.
